core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java

```diff
@@ -271,25 +277,47 @@ public void process(StreamDecoderInput in, StreamDecoderOutput<WebSocketFrame> o
                         continue; // to while loop
                     }
 
+                    if (frameOpcode != WebSocketFrameType.TEXT.opcode() &&
+                        frameOpcode != WebSocketFrameType.BINARY.opcode() &&
+                        frameOpcode != WebSocketFrameType.CONTINUATION.opcode()) {
+                        throw protocolViolation(WebSocketCloseStatus.INVALID_MESSAGE_TYPE,
+                                                "Cannot decode a web socket frame with opcode: " + frameOpcode);
+                    }
+
                     final WebSocketFrame decodedFrame;
                     if (frameOpcode == WebSocketFrameType.TEXT.opcode()) {
                         decodedFrame = WebSocketFrame.ofPooledText(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
                     } else if (frameOpcode == WebSocketFrameType.BINARY.opcode()) {
                         decodedFrame = WebSocketFrame.ofPooledBinary(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
-                    } else if (frameOpcode == WebSocketFrameType.CONTINUATION.opcode()) {
-                        decodedFrame = WebSocketFrame.ofPooledContinuation(payloadBuffer, finalFragment);
-                        out.add(decodedFrame);
                     } else {
-                        throw protocolViolation(WebSocketCloseStatus.INVALID_MESSAGE_TYPE,
-                                                "Cannot decode a web socket frame with opcode: " + frameOpcode);
+                        assert frameOpcode == WebSocketFrameType.CONTINUATION.opcode();
+                        decodedFrame = WebSocketFrame.ofPooledContinuation(payloadBuffer, finalFragment);
                     }
                     logger.trace("{} is decoded.", decodedFrame);
+
                     if (finalFragment) {
                         fragmentedFramesCount = 0;
```

Comment:
Side-note) This comment is not related to this PR. I don't see we need to track `fragmentedFramesCount` based on the current usage. a boolean flag seems enough to know fragmented frames are received.

Reply:
That is correct. When I forked the original code, I also noticed that `fragmentedFramesCount` can be a boolean field but I didn't change it because
- There's no not so much performance penalty
- Keeping the original code as much as possible makes it simpler to update the forked source code later if there are changes in the upstream code.

https://github.com/netty/netty/blob/e8df52e442629214e0355528c00e873e213f0139/codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocket08FrameDecoder.java#L98

Reply:
Understood. It would be better to send a PR first before modifying it. Let's keep it as is for now.